ua-snap/prefect
# Regridding via Prefect

This PR closes #16 and is a companion to [PR #40](https://github.com/ua-snap/cmip6-utils/pull/40) for the `regrid_prefect` [branch of the cmip6-utils repo](https://github.com/ua-snap/cmip6-utils/tree/regrid_prefect).

Using the indicators work as a template, this branch adds a Prefect flow that orchestrates the regridding pipeline. A bunch of code is reused from the indicators work (mostly the setup tasks and the tasks to monitor slurm jobs). 

Along with the typical parameters for user info and scratch directory, the user has an option here to skip generating batch files (useful to save 30 minutes in the flow run as we keep developing the regridding pipeline on the same set of input files) and to prevent overwriting previously regridded files (via the `no_clobber` parameter). _(Note that the `no_clobber` option is not actually implemented yet, but the parameter is included here and has been passed thru to the appropriate functions!)_

Some changes were made to the regridding pipeline in order to accomplish this, and those changes are in the [regrid_prefect branch](https://github.com/ua-snap/cmip6-utils/tree/regrid_prefect) of the `cmip6-utils` repo. Those changes are breifly described here:

- Instead of using `config.py` and environmental variables, the various directories needed for the pipeline are just constructed from the `scratch_directory` parameter in the `regrid_cmip6.regrid_cmip6()` function. (This is similar to what we did in the inidicators flow.) As a result, more arguments have to be passed to `slurm.py` and `regrid.py` for all the directories since they are not readily available from `config.py`.
- A new `run_generate_batch_files.py` script was needed to pass parameters to `generate_batch_files.py`, create the sbatch head and slurm file, and submit a the slurm job for that process. 
-  The step of submitting the slurm jobs was moved out of the `regrid_cmip6.ipynb` notebook and into the main block of `slurm.py`

**TO TEST:**

- On the `regridding` branch of this repo, start the server and run ```python regrid_cmip6.py```
- Start a new flow run using the following parameters (substituting your ssh/user info, email, and your own scratch dir):
```
{
  "ssh_username": "jdpaul3",
  "ssh_private_key_path": "/Users/joshpaul/.ssh/id_rsa",
  "branch_name": "regrid_prefect",
  "cmip6_directory": "/beegfs/CMIP6/arctic-cmip6/CMIP6",
  "scratch_directory": "/center1/CMIP6/jdpaul3/regrid_prefect",
  "slurm_email": "jdpaul3@alaska.edu",
  "no_clobber": true,
  "generate_batch_files": true
}
```
- Wait until you are sure that the `run_generate_batch_files()` task has completed, and the `create_and_run_slurm_scripts()` task has started. At this point you can cancel the flow run and also cancel the slurm jobs from your terminal using ```scancel -u your_chinook_username```
- Check out the batch files that were created (there should be 260 of them).
- Delete all but a random handful of the batch files. I ran the following in the `{scratch_directory}/regrid_batch` directory:
```
mkdir tmp
mv -t ./tmp batch_EC-Earth3-Veg_ssp245_gr0_13.txt batch_KACE-1-0-G_historical_gr1_0.txt batch_CNRM-CM6-1-HR_ssp126_gr0_0.txt batch_TaiESM1_ssp585_gr1_0.txt 
rm *.txt
mv tmp/* .
rm -rf tmp
```
- Now start a brand new prefect flow, but change the `generate_batch_files` parameter input to "false". Run the flow, and you should see that the `run_generate_batch_files()` task has been skipped and the `create_and_run_slurm_scripts()` task should start immediately after the setup tasks are done. 
- Wait for the process to complete and verify that regridded files exist in the `{scratch_directory}/regrid` directory
